### PR TITLE
MAGECLOUD-3251: Add Interactive Service Compatibility Validation

### PR DESCRIPTION
**Pull request by NadiyaS**
### Description
Add Interactive Service Compatibility Validation

### Fixed Issues (if relevant)
https://magento2.atlassian.net/browse/MAGECLOUD-3251

### Manual testing scenarios

https://jira.corp.magento.com/browse/MAGECLOUD-50

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/ece-tools?pullRequest=445) <br/>All committers have signed the CLA.

File: src/Docker/Service/Version/Validator.php [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+     * @var array
+     */
+    private $supportedVersionList;
+
+    /**
+     * @param MagentoVersion $magentoVersion
+     */
+    public function __construct(MagentoVersion $magentoVersion)
+    {
+
```
Comment (oshmyheliuk) [inline src/Docker/Service/Version/Validator.php]: Please remove empty line.

File: src/Docker/Service/Config.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    const KEY_CRON = 'cron';
+    const KEY_VARNISH = 'varnish';
+
+    /**
+     * List of services which can be configured in Cloud docker
+     *
+     * @var array
+     */
+    private $configurableVersions = [
```
Comment (oshmyheliuk) [inline src/Docker/Service/Config.php]: I think is better to name this variable as `$configurableServices`

File: src/Docker/Service/Config.php [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+     * Example of return:
+     *
+     * ```php
+     *  [
+     *      'elasticserach' => '5.6',
+     *      'db' => '10.0'
+     *  ];
+     * ```
+     *
+     * @param Repository $customVersions overwrite version from config files
```
Comment (oshmyheliuk) [inline src/Docker/Service/Config.php]: `overwrite version from config files` looks like description isn't correct as ` $customVersions` contains services versions from `docker:build` command input.

File: src/Docker/Service/Config.php [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        }
+
+        /**
+         * We don't support release candidates.
+         */
+        return rtrim($version, '-rc');
+    }
+
+    /**
+     * @return array
```
Comment (oshmyheliuk) [inline src/Docker/Service/Config.php:137]: Add a description to the method

File: src/Docker/Service/Version/Validator.php (hunk 1 above)
Comment (NadiyaS) [reply src/Docker/Service/Version/Validator.php]: Fixed

File: src/Docker/Service/Config.php (hunk 2 above)
Comment (NadiyaS) [reply src/Docker/Service/Config.php]: agree. Fixed

File: src/Docker/Service/Config.php (hunk 3 above)
Comment (NadiyaS) [reply src/Docker/Service/Config.php]: I meant that value from this argument will overwrite versions set in config files. 
is `custom version which overwrite values from configuration files` be more clear?

File: src/Docker/Service/Version/Validator.php [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * Validates provided services version for current Magento.
+     * Returns empty array if all provided versions are supported. Otherwise, returns warning message for
+     * every unsupported service in separate array elements.
+     *
+     * Example of $serviceVersions argument:
+     *
+     * ```php
+     *  [
+     *      'elasticserach' => '5.6',
```
Comment (YPyltiai) [inline src/Docker/Service/Version/Validator.php]: Miss-spelling here. `elasticsearch`

**Comment by YPyltiai** [comment]
I may be missing something but looks like we miss this AC:
```
Add no-interaction / force flag, so CICD processes can skip this in case they need potentially unsupported stack with installed Magento version or ECE-Tools mapping is behind
```

File: src/Docker/Service/Version/Validator.php (hunk 5 above)
Comment (NadiyaS) [reply src/Docker/Service/Version/Validator.php]: thanks, fixed

**Comment by NadiyaS** [comment]
@YPyltiai we already have built-in option for such cases **-n, --no-interaction** 
Should we add additional option -f which will duplicate this behavior?

**Comment by YPyltiai** [comment]
> @YPyltiai we already have built-in option for such cases **-n, --no-interaction**
> Should we add additional option -f which will duplicate this behavior?

I didn't know we have it for all commands, while also didn't find it in current scope of changes. If it works - that's fine. Just another thing for @andriyShevtsov and @mveeramneni to validate during QA phase then.

File: src/Docker/Service/Version/Validator.php [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+            Config::KEY_ELASTICSEARCH => '1.7 || ^2.0',
+            Config::KEY_RABBITMQ => '~3.5',
+        ],
+        '>=2.2.0 <2.2.8' => [
+            Config::KEY_PHP => '~7.0.13 || ~7.1',
+            Config::KEY_DB => '>=10.0 <10.3',
+            Config::KEY_NGINX => '^1.0',
+            Config::KEY_VARNISH => '~4.0 || ~5.0',
+            Config::KEY_REDIS => '~3.2 || ~4.0 || ~5.0',
+            Config::KEY_ELASTICSEARCH => '^2.0 || ^5.0',
```
Comment (YPyltiai) [inline src/Docker/Service/Version/Validator.php]: @NadiyaS , also had a question why we use whole major release line in constraints. Currently cloud supports separate defined minor versions of services. For example - for 5.x release line it's 5.2. Not sure if that's fine to leave wide constraints, while all other services are defined in a more accurate way (like MariaDB / Redis, etc.)

File: src/Docker/Service/Config.php (hunk 3 above)
Comment (oshmyheliuk) [reply src/Docker/Service/Config.php]: yes, thank you

File: src/Docker/Service/Config.php (hunk 4 above)
Comment (NadiyaS) [reply src/Docker/Service/Config.php:137]: added

File: src/Docker/Service/Version/Validator.php (hunk 6 above)
Comment (NadiyaS) [reply src/Docker/Service/Version/Validator.php]: Discussed offline. Versions were changed.

File: src/Docker/Service/Version/Validator.php [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+     * Validates service version whether it is supported by current Magento version or not.
+     *
+     * @param string $serviceName Service name
+     * @param string $version Service version for validation
+     * @return string Failed validation message
+     *
+     * @throws ConfigurationMismatchException
+     * @throws UndefinedPackageException
+     */
+    private function validateService($serviceName, $version)
```
Comment (oshmyheliuk) [inline src/Docker/Service/Version/Validator.php]: Please add argument and return types

File: src/Docker/Service/Version/Validator.php [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+     *  ];
+     * ```
+     *
+     * @param array $serviceVersions List of services and their names which should be validates.
+     * @return array List of warning messages. One message for one unsupported service.
+     *
+     * @throws ConfigurationMismatchException
+     * @throws UndefinedPackageException
+     */
+    public function validateVersions($serviceVersions)
```
Comment (oshmyheliuk) [inline src/Docker/Service/Version/Validator.php]: Please add argument and return types

**Review by oshmyheliuk — APPROVED** [review]

**Review by YPyltiai — APPROVED** [review]
Looks good now. @shiftedreality , please check that constraints match all docker image versions, please.

**Review by shiftedreality — APPROVED** [review]

**Review by shiftedreality — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request closed without merging**